packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1428 @@
+                uint128 sharesToBurn = delegator.undelegateRequests[requestTimes[i]];
+
+                unchecked {
+                    assets += sharesToBurn.mulDiv(
+                        _vault[validator].asset.totalPendingAssets,
+                        _vault[validator].asset.totalPendingKroShares
+                    );
+                    shares += sharesToBurn;
+                }
+
+                delete delegator.undelegateRequests[requestTimes[i]];
+                delete delegator.undelegateRequestTimes[i];
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        _vault[validator].asset.totalPendingAssets -= assets;
+        _vault[validator].asset.totalPendingKroShares -= shares;
+
+        bool success = IERC20(TOKEN).transfer(msg.sender, assets);
+        require(success, "AssetManager: Transfer failed");
+
+        emit KroUndelegationFinalized(validator, msg.sender, assets);
+        return assets;
+    }
+
+    /**
+     * @notice Finalize all pending KGH undelegation and returns the amount of assets that the vault would
+     *         exchange for the amount of shares provided.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The amount of assets that the vault would exchange for the amount of shares provided.
+     */
+    function finalizeUndelegateKgh(address validator) external returns (uint128) {
+        KghDelegator storage kghDelegator = _vault[validator].kghDelegators[msg.sender];
+        uint256[] memory requestTimes = kghDelegator.undelegateRequestTimes;
+        uint128 assets = finalizeUndelegate(validator);
+        uint128 shares;
+
+        for (uint256 i = 0; i < requestTimes.length; ) {
+            if (requestTimes[i] + WITHDRAWAL_DELAY <= block.timestamp) {
+                uint128 kghSharesToBurn = kghDelegator.undelegateRequests[requestTimes[i]];
+                uint256[] memory tokenId = kghDelegator.undelegateRequestedIds[requestTimes[i]];
+
+                if (_vault[validator].asset.totalPendingKghShares != 0) {
+                    unchecked {
+                        assets += kghSharesToBurn.mulDiv(
+                            _vault[validator].asset.totalPendingBoostedRewards,
+                            _vault[validator].asset.totalPendingKghShares
+                        );
+                        shares += kghSharesToBurn;
+                    }
+                }
+
+                for (uint256 tokenIdIndex = 0; tokenIdIndex < tokenId.length; ) {
+                    KGH.safeTransferFrom(address(this), msg.sender, tokenId[tokenIdIndex]);
+
+                    unchecked {
+                        ++tokenIdIndex;
+                    }
+                }
+
+                delete kghDelegator.undelegateRequests[requestTimes[i]];
+                delete kghDelegator.undelegateRequestTimes[i];
+            }
+
+            unchecked {
+                ++i;
+            }
```

Comment:
If the `undelegateRequestTimes` is pushed orderly, we can break the for loop earlier.
```suggestion
                unchecked {
                    ++i;
                }
            } else {
                break;
            }
```